File: docs/_downloads/1cd84d2dcc8bc3927913e970976e9072/Q784.java
package leet.java.medium;

import java.util.ArrayList;
import java.util.List;

/**<a href='https://leetcode.com/problems/letter-case-permutation/'>
 * 784. Letter Case Permutation</a>
 * 
 * @author Odys Zhou
 */
public class Q784 {
    public List<String> letterCasePermutation(String S) {
    	return btrack(S);
    }

	/** 49.34% 
	 * @param s
	 * @return
	 */
	private List<String> btrack(String s) {
    	List<String> res = new ArrayList<String>();
    	int l = s.length();
    	if (l == 0) {
    		res.add("");
    		return res;
    	}
    	String last = s.substring(l - 1, l);
    	List<String> tracks = btrack(s.substring(0, s.length() - 1));
		boolean isdigit = false;
		try {
			Integer.valueOf(last);
			isdigit = true;
		} catch(Exception e) {}
    	for (String t : tracks) {
    		if (isdigit)
    			res.add(t + last);
    		else {
    			res.add(t + last.toLowerCase());
    			res.add(t + last.toUpperCase());
    		}
    	}
		return res;
	}


}

File: docs/_downloads/414531146a8ac482e05ce04ab1710685/q784.py
'''
784. Letter Case Permutation
https://leetcode.com/problems/letter-case-permutation/

Given a string S, we can transform every letter individually to be lowercase or uppercase to create
another string.

Return a list of all possible strings we could create. You can return the output in any order.

Example 1:
Input: S = "a1b2"
Output: ["a1b2","a1B2","A1b2","A1B2"]

Example 2:
Input: S = "3z4"
Output: ["3z4","3Z4"]

Example 3:
Input: S = "12345"
Output: ["12345"]

Example 4:
Input: S = "0"
Output: ["0"]
 

Constraints:

S will be a string with length between 1 and 12.
S will consist only of letters or digits.

Created on 23 Feb 2021

@author: Odys Zhou
'''
from unittest import TestCase
from typing import List

class Solution:
    ''' 88.03%
    '''
    def letterCasePermutation(self, S: str) -> List[str]:
        def btracking(s: str) -> List[str]:
            if len(s) == 0: return []
            elif len(s) == 1: return [s] if s.isdigit() else [s.lower(), s.upper()]

            res = []
            tracks = btracking(s[:-1])
            for t in tracks:
                if (s[-1].isdigit()):
                    res.append(t + s[-1])
                else:
                    res.append(t + s[-1].lower())
                    res.append(t + s[-1].upper())
            return res
        
        return btracking(S)

if __name__ == '__main__':
    t = TestCase()
    s = Solution()

    t.assertCountEqual(['a1b2', 'a1B2', 'A1b2', 'A1B2'], s.letterCasePermutation('a1b2'))
    t.assertCountEqual(['3z4', '3Z4'], s.letterCasePermutation('3z4'))
    t.assertCountEqual(['3z4', '3Z4'], s.letterCasePermutation('3Z4'))
    t.assertCountEqual(['12345'], s.letterCasePermutation('12345'))
    t.assertCountEqual(['0'], s.letterCasePermutation('0'))
    t.assertCountEqual(['a', 'A'], s.letterCasePermutation('A'))
    
    print('OK!')